docs/tools/testing/unit_testing.md

```diff
@@ -1,27 +1,275 @@
 ## Unit testing
 
-We perform integration and system testing in an environment where we run the tests with the full Mbed OS. Other testing tools for Mbed OS require specific embedded hardware, which means traditional unit testing is not possible.
+This document describes how to run and write unit tests for Mbed OS. In case of running into problems see [troubleshooting](#troubleshooting) section.
+
+### Introduction
+
+We perform integration and system testing in an environment where we run the tests with the full Mbed OS. Other testing tools for Mbed OS require specific embedded hardware, which means actual unit testing is not possible.
 
 Unit testing takes place in a build environment where we test each C or C++ class or module in isolation. This means we build test suites into separate test binaries and stub all access outside to remove dependencies on any specific embedded hardware or software combination. This allows us to complete the testing using native compilers on the build machine.
 
-### Using unit tests
+### Prerequisites
+
+* GNU toolchains installed.
+    * GCC 6 or later
+    * We recommend you use MinGW-W64 on Windows, but any Windows port of the above GCC versions works.
+* CMake 3.0+ installed.
+* Python 2.7.x or >3.5 and pip 10.0 (or newer) installed.
+* Gcovr >=4.1
+* Mbed CLI 1.8.0 or later.
+
+#### Installing dependencies on Debian/Ubuntu
+
+1. `sudo apt-get -y install build-essential cmake`
+2. Install python and pip:
+```
+sudo apt-get -y install python python-setuptools
+sudo easy_install pip
+```
+3. Install [gcovr](#installing-covr).
+4. (Optional) Install [Mbed CLI](https://os.mbed.com/docs/latest/tools/arm-mbed-cli.html).
+
+#### Installing dependencies on Mac OS
+
+1. Install [Homebrew](https://brew.sh/).
+2. Install gcc compilers and cmake with: `brew install gcc cmake`
+3. Install python and pip:
+```
+brew install python
+sudo easy_install pip
+```
+3. Install [gcovr](#installing-covr).
+4. (Optional) Install [Mbed CLI](https://os.mbed.com/docs/latest/tools/arm-mbed-cli.html).
+
+#### Installing dependencies on Windows
+
+1. Download and install [MinGW-W64](http://mingw-w64.org/).
+2. Download CMake binaries from https://cmake.org/download/ and run the installer.
+3. Download Python2.7 or Python3 from https://www.python.org/getit/ and run the installer.
+4. Add MinGW, CMake and Python into PATH.
+5. Install [gcovr](#installing-covr).
+6. (Optional) Install [Mbed CLI](https://os.mbed.com/docs/latest/tools/arm-mbed-cli.html).
+
+#### Installing gcovr
+
+Install gcovr code coverage tool globally with `pip install 'gcovr>=4.1'` or if you need an isolated Python environment use virtualenv:
```

Comment:
Why are you explaining what it is just now (code coverage tool)? Explain that in the prereq list where you first told me I need it.